api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirLocationDaoImpl.java

```diff
@@ -44,35 +53,93 @@ public Location getLocationByUuid(String uuid) {
 		return locationService.getLocationByUuid(uuid);
 	}
 	
-	@Override
-	public Collection<Location> findLocationByName(String name) {
-		return locationService.getLocations(name);
+	private void handleName(Criteria criteria, StringParam namePattern) {
+		if (namePattern != null) {
+			propertyLike("name", namePattern).ifPresent(criteria::add);
+		}
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByCity(String city) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("cityVillage", city)).list();
+	private void handleCity(Criteria criteria, StringParam city) {
+		if (city != null) {
+			propertyLike("cityVillage", city).ifPresent(criteria::add);
+		}
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByCountry(String country) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("country", country)).list();
+	private void handleCountry(Criteria criteria, StringParam country) {
+		if (country != null) {
+			propertyLike("country", country).ifPresent(criteria::add);
+		}
 	}
 	
-	@Override
-	public Collection<Location> findLocationsByPostalCode(String postalCode) {
-		return sessionFactory.getCurrentSession().createCriteria(Location.class).add(eq("postalCode", postalCode)).list();
+	private void handlePostalCode(Criteria criteria, StringParam postalCode) {
+		if (postalCode != null) {
+			propertyLike("postalCode", postalCode).ifPresent(criteria::add);
+		}
+	}
+	
+	private void handleState(Criteria criteria, StringParam state) {
+		if (state != null) {
+			propertyLike("stateProvince", state).ifPresent(criteria::add);
+		}
+	}
+	
+	private void handleTag(Criteria criteria, TokenParam tag) {
+		if (tag != null) {
+			criteria.createAlias("tags", "t");
+			propertyLike("t.name", tag).ifPresent(criteria::add);
+		}
+	}
+	
+	// handle queries of the form /Location?partof=parent_name
+	private void handleParentLocation(Criteria criteria, StringParam parent) {
+		if (parent != null) {
+			ProjectionList projList = Projections.projectionList();
+			projList.add(Projections.property("locationId"));
+			
+			Criteria criteriaForParent = this.sessionFactory.getCurrentSession().createCriteria(Location.class);
+			propertyLike("name", parent).ifPresent(criteriaForParent::add);
+			
+			Set<Object> possibleParentLocationIdList = new HashSet<>(criteriaForParent.setProjection(projList).list());
```

Comment:
Instead of this:

```java
criteriaForParent.setProjection(projectionList);

criteria.add(Subqueries.in("parent.locationId", criteriaForParent));
```

Your current implementation results in 2 db queries. This results in just one, which should be our goal.

Reply:
I was stuck on this for a while because we need to use "Subqueries.propertyIn" and not "Subqueries.in" as the test was failing repeatedly. Thanks for the hint though :)